### PR TITLE
Fix httpRequestUri and populate body

### LINKED ISSUE #6305: [Custom Handlers]Forward HttpTrigger request when enableForwardingHttpRequest=true
The handler is always invoked with the function's name as the path.

Looks like the problem is here:
https://github.com/Azure/azure-functions-host/blob/242d08327b9cb751d7f9ccae1fb8d4570c8b0286/src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs#L175

`httpRequestMessage` is always empty because:
https://github.com/Azure/azure-functions-host/blob/242d08327b9cb751d7f9ccae1fb8d4570c8b0286/src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs#L161-L162

### PR DESCRIPTION
**Pull request by yojagad**
### Issue describing the changes in this PR

Fixes #6305

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [x] Otherwise: Backport tracked by issue/PR #6333 
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additionally ran E2E tests with Go samples from https://github.com/Azure-Samples/functions-custom-handlers

### CODE REVIEW

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 1]
```diff
@@ -69,8 +69,19 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
...
+                    httpRequestMessage.Content = new StreamContent(httpRequest.Body);
+
+                    if (!string.IsNullOrEmpty(httpRequest.ContentType))
+                    {
+                        httpRequestMessage.Content.Headers.Add("Content-Type", httpRequest.ContentType);
+                    }
+                    if (httpRequest.ContentLength != null)
+                    {
+                        httpRequestMessage.Content.Headers.Add("Content-Length", httpRequest.ContentLength.ToString());
+                    }
```
Comment (anthonychu) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Does this mean other headers will be stripped out? I think all headers should be forwarded.

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Ideally we need to avoid rebuilding each part in http request when forwarding. Using `HttpRequestMessageFeature` gave us that ability. Http request will be forwarded and only additional headers from host will be added.

File: src/WebJobs.Script/Extensions/HttpRequestExtensions.cs [hunk 2]
```diff
@@ -122,6 +124,35 @@ public static bool IsMediaTypeOctetOrMultipart(this HttpRequest request)
             return false;
         }
 
+        public static Task<HttpRequestMessage> GetProxyHttpRequest(this HttpRequest request, string requestUri, string invocationId)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Extensions/HttpRequestExtensions.cs]: rename to GetHttpRequestMessage

Comment (pragnagopa) [inline src/WebJobs.Script/Extensions/HttpRequestExtensions.cs]: Unit tests for this? Or is this covered in DefaultHttpWorkerService tests?

File: src/WebJobs.Script/Extensions/HttpRequestExtensions.cs [hunk 3]
```diff
@@ -122,6 +124,35 @@ public static bool IsMediaTypeOctetOrMultipart(this HttpRequest request)
...
+        {
+            HttpRequestMessage proxyRequest = new HttpRequestMessage();
+            proxyRequest.RequestUri = new Uri(QueryHelpers.AddQueryString(requestUri, request.GetQueryCollectionAsDictionary()));
+
+            foreach (var header in request.Headers)
+            {
+                proxyRequest.Headers.TryAddWithoutValidation(header.Key, header.Value.FirstOrDefault());
+            }
+            proxyRequest.Headers.Add(HttpWorkerConstants.HostVersionHeaderName, ScriptHost.Version);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Extensions/HttpRequestExtensions.cs]: Header specific to HttpWorker should be in `DefaultHttpWorkerService`. HttpRequest extension is generic and should not contain custom handler specific headers to allow other parts for the system to use this method

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 4]
```diff
@@ -69,7 +67,7 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
 
             try
             {
-                using (HttpRequestMessage httpRequestMessage = CreateAndGetHttpRequestMessage(scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId.ToString(), new HttpMethod(httpRequest.Method), httpRequest.GetQueryCollectionAsDictionary()))
+                using (HttpRequestMessage httpRequestMessage = await httpRequest.GetProxyHttpRequest(BuildAndGetUri(GetPathValue(_httpWorkerOptions, scriptInvocationContext.FunctionMetadata.Name, httpRequest)), scriptInvocationContext.ExecutionContext.InvocationId.ToString()))
                 {
                     _logger.LogDebug("Sending http request message for simple httpTrigger function: '{functionName}' invocationId: '{invocationId}'", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Nit: Update log to `Forwarding Http request for HttpTrigger` ? Remove `Simple`

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 5]
```diff
@@ -69,7 +67,7 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
 
             try
             {
-                using (HttpRequestMessage httpRequestMessage = CreateAndGetHttpRequestMessage(scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId.ToString(), new HttpMethod(httpRequest.Method), httpRequest.GetQueryCollectionAsDictionary()))
+                using (HttpRequestMessage httpRequestMessage = await httpRequest.GetProxyHttpRequest(BuildAndGetUri(GetPathValue(_httpWorkerOptions, scriptInvocationContext.FunctionMetadata.Name, httpRequest)), scriptInvocationContext.ExecutionContext.InvocationId.ToString()))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Nit: Please nested function calls and call them explicitly to improve readability

Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Nit: In BuildAndGetUri method, flip if condition to check for null or empty first. Improves readability.

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 6]
```diff
@@ -91,15 +89,26 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
             }
         }
 
+        internal string GetPathValue(HttpWorkerOptions httpWorkerOptions, string functionName, HttpRequest httpRequest)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs:105]: Add unit tests for `BuildAndGetUri(string pathValue = null)` method

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 7]
```diff
@@ -91,15 +89,26 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
...
+        }
+
         internal async Task ProcessDefaultInvocationRequest(ScriptInvocationContext scriptInvocationContext)
         {
             try
             {
                 HttpScriptInvocationContext httpScriptInvocationContext = await scriptInvocationContext.ToHttpScriptInvocationContext();
                 // Build httpRequestMessage from scriptInvocationContext
-                using (HttpRequestMessage httpRequestMessage = CreateAndGetHttpRequestMessage(scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId.ToString(), HttpMethod.Post))
+
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs:122]: Nit: remove empty line after comment

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 8]
```diff
@@ -91,15 +89,26 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
...
+                using (HttpRequestMessage httpRequestMessage = await scriptInvocationContext.ToProxyHttpRequest(BuildAndGetUri(scriptInvocationContext.FunctionMetadata.Name), httpScriptInvocationContext))
                 {
-                    httpRequestMessage.Content = new ObjectContent<HttpScriptInvocationContext>(httpScriptInvocationContext, new JsonMediaTypeFormatter());
                     _logger.LogDebug("Sending http request for function:{functionName} invocationId:{invocationId}", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
                     HttpResponseMessage response = await _httpClient.SendAsync(httpRequestMessage);
                     _logger.LogDebug("Received http request for function:{functionName} invocationId:{invocationId}", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Nit: Fix log `Received http response`

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 9]
```diff
@@ -91,15 +89,26 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
...
+
         internal async Task ProcessDefaultInvocationRequest(ScriptInvocationContext scriptInvocationContext)
         {
             try
             {
                 HttpScriptInvocationContext httpScriptInvocationContext = await scriptInvocationContext.ToHttpScriptInvocationContext();
                 // Build httpRequestMessage from scriptInvocationContext
-                using (HttpRequestMessage httpRequestMessage = CreateAndGetHttpRequestMessage(scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId.ToString(), HttpMethod.Post))
+
+                using (HttpRequestMessage httpRequestMessage = await scriptInvocationContext.ToProxyHttpRequest(BuildAndGetUri(scriptInvocationContext.FunctionMetadata.Name), httpScriptInvocationContext))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Same here. Remove nested function calls for readability.

File: src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs [hunk 10]
```diff
@@ -104,6 +106,22 @@ private static bool ShouldSkipBindingData(KeyValuePair<string, object> bindingDa
             return false;
         }
 
+        public static Task<HttpRequestMessage> ToProxyHttpRequest(this ScriptInvocationContext context, string requestUri, HttpScriptInvocationContext content)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: Reaname to `ToHttpRequestMessage`

File: src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs [hunk 11]
```diff
@@ -104,6 +106,22 @@ private static bool ShouldSkipBindingData(KeyValuePair<string, object> bindingDa
...
+        {
+            HttpRequestMessage proxyRequest = new HttpRequestMessage();
+            proxyRequest.RequestUri = new Uri(requestUri);
+
+            proxyRequest.Headers.Add(HttpWorkerConstants.HostVersionHeaderName, ScriptHost.Version);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: Same here. These headers are specific to http worker. Move them to a helper within in `DefaultHttpWorkerService` and reuse that help in both cases forwarding and non-forwarding

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 12]
```diff
@@ -69,7 +67,7 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
 
             try
             {
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs:69]: Please add null checks for logger, httpclient and httpworkeroptions in constructor. I missed this earlier.

File: test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs [hunk 13]
```diff
@@ -33,6 +33,8 @@ public static HttpRequest GetTestHttpRequest()
         {
             var httpRequest = new DefaultHttpContext().Request;
             httpRequest.Method = "GET";
+            httpRequest.Scheme = "http";
+            httpRequest.Host = new HostString("localhost", 8080);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs]: Is it possible to take input string for hoststring? that will allow us to add more tests on request path. If too many changes we can skip it.

**Review by pragnagopa — COMMENTED** [review]
Added comments. Thanks!

File: src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs (hunk 11 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: So, if I do remove the headers from here, there's actually nothing I use from ScriptInvocationContext to build the HttpRequestMessage. Therefore, I'm not sure if it makes sense to expose it as an extension at all ?

Comment (yojagad) [reply src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: I might makes sense to expose it as HttpScriptInvocationContextExtension though, let me know what you think

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: It only adding headers, lets skip extension method. Use helpers instead.

Comment (yojagad) [reply src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: It also sets the body from HttpScriptInvocationContext which is why I was suggesting going down the HttpScriptInvocationContextExtension path...still stick with helper approach?

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs (hunk 1 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: All headers are forwarded now.

File: src/WebJobs.Script/Extensions/HttpRequestExtensions.cs (hunk 2 above)
Comment (yojagad) [reply src/WebJobs.Script/Extensions/HttpRequestExtensions.cs]: Added more tests

File: src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs (hunk 11 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: Changed to HttpScriptInvocationContextExtension as discussed

File: test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs (hunk 13 above)
Comment (yojagad) [reply test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs]: Yeah, added input string

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_CustomHandler.cs [hunk 14]
```diff
@@ -52,7 +52,7 @@ private async Task InvokeHttpTrigger(string functionName)
 
             string responseContent = await response.Content.ReadAsStringAsync();
             JObject res = JObject.Parse(responseContent);
-            Assert.True(res["functionName"].ToString().StartsWith(functionName));
+            Assert.True(res["functionName"].ToString().StartsWith($"api/{functionName}"));
```
Comment (yojagad) [inline test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_CustomHandler.cs:55]: @pragnagopa Wanted to double check this. Since it's custom handler and EnableRequestForwarding was set to true in this scenario, this had the value of "api/functionName"
.

File: src/WebJobs.Script/Extensions/HttpRequestExtensions.cs [hunk 15]
```diff
@@ -122,6 +123,34 @@ public static bool IsMediaTypeOctetOrMultipart(this HttpRequest request)
             return false;
         }
 
+        public static HttpRequestMessage GetHttpRequestMessage(this HttpRequest request, string requestUri)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Extensions/HttpRequestExtensions.cs]: Rename method to `ToHttpRequestMessage`

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs [hunk 16]
```diff
@@ -69,11 +67,17 @@ internal async Task ProcessHttpInAndOutInvocationRequest(ScriptInvocationContext
...
+                string uri = BuildAndGetUri(uriPathValue);
+
+                using (HttpRequestMessage httpRequestMessage = httpRequest.GetHttpRequestMessage(uri))
                 {
-                    _logger.LogDebug("Sending http request message for simple httpTrigger function: '{functionName}' invocationId: '{invocationId}'", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
+                    AddHeaders(httpRequestMessage, scriptInvocationContext.ExecutionContext.InvocationId.ToString());
+
+                    _logger.LogDebug("Forwarding http request for httpTrigger function: '{functionName}' invocationId: '{invocationId}'", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
                     HttpResponseMessage invocationResponse = await _httpClient.SendAsync(httpRequestMessage);
                     _logger.LogDebug("Received http response for simple httpTrigger function: '{functionName}' invocationId: '{invocationId}'", scriptInvocationContext.FunctionMetadata.Name, scriptInvocationContext.ExecutionContext.InvocationId);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs]: Nit: remove simple for 'Received http response for simple httpTrigger function'

File: src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs [hunk 17]
```diff
@@ -4,6 +4,8 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Formatting;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/MessageExtensions/ScriptInvocationContextExtensions.cs]: undo

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_CustomHandler.cs (hunk 14 above)
Comment (pragnagopa) [reply test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_CustomHandler.cs:55]: Correct! Thanks!

File: test/WebJobs.Script.Tests/HttpWorker/DefaultHttpWorkerServiceTests.cs [hunk 18]
```diff
@@ -276,6 +277,40 @@ public async Task ProcessSimpleHttpTriggerInvocationRequest_CustomHandler_Enable
...
+            HttpWorkerOptions testOptions = new HttpWorkerOptions
+            {
+                Port = 8080,
+            };
+
+            if (string.IsNullOrEmpty(pathValue))
+            {
+                expectedUriString = new UriBuilder(WorkerConstants.HttpScheme, WorkerConstants.HostName, testOptions.Port).ToString();
+            }
+            expectedUriString = new UriBuilder(WorkerConstants.HttpScheme, WorkerConstants.HostName, testOptions.Port, pathValue).ToString();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/HttpWorker/DefaultHttpWorkerServiceTests.cs]: instead of building it here, pass this as input to the test

File: test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs [hunk 19]
```diff
@@ -29,10 +29,12 @@ public class HttpWorkerTestUtilities
         public const string HttpResponseContentStringValue = "hello world response";
         public const string HttpResponseNotFoundValue = "Route not found";
 
-        public static HttpRequest GetTestHttpRequest()
+        public static HttpRequest GetTestHttpRequest(string host = "localhost")
         {
             var httpRequest = new DefaultHttpContext().Request;
             httpRequest.Method = "GET";
+            httpRequest.Scheme = "http";
+            httpRequest.Host = new HostString(host, 8080);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/HttpWorker/HttpWorkerTestUtilities.cs]: change to `httpRequest.Host = new HostString(host);`

File: test/WebJobs.Script.Tests/HttpWorker/DefaultHttpWorkerServiceTests.cs [hunk 20]
```diff
@@ -393,6 +428,24 @@ public void ProcessOutputLogs_Succeeds(HttpScriptInvocationResult httpScriptInvo
...
         }
 
+        [Theory]
+        [InlineData("someFuntionName", CustomHandlerType.Http, true, "/")]
+        [InlineData("someFuntionName", CustomHandlerType.Http, false, "someFuntionName")]
+        [InlineData("someFuntionName", CustomHandlerType.None, true, "someFuntionName")]
+        [InlineData("someFuntionName", CustomHandlerType.None, false, "someFuntionName")]
+        public void TestPathValue(string functionName, CustomHandlerType type, bool enableForwardingHttpRequest, string expectedValue)
+        {
+            HttpRequest testHttpRequest = HttpWorkerTestUtilities.GetTestHttpRequest();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/HttpWorker/DefaultHttpWorkerServiceTests.cs]: you can now test different paths for example : ` HttpRequest testHttpRequest = HttpWorkerTestUtilities.GetTestHttpRequest("localhost:8000/api/foo/duh");` , take host string as input

**Review by pragnagopa — COMMENTED** [review]
Added few more. Mostly minor nits. Thanks.

**Comment by yojagad** [comment]
Addressed the additional comments.

**Review by pragnagopa — APPROVED** [review]
👍 Please update description with E2E tests verified from https://github.com/Azure-Samples/functions-custom-handlers
